lib/src/cli/flutter_cli.dart

```diff
@@ -65,55 +62,182 @@ class Flutter {
   static Future<void> test({
     String cwd = '.',
     bool recursive = false,
-    void Function([String?]) Function(String message)? progress,
-  }) async {
-    await _runCommand(
-      cmd: (cwd) async {
-        final installDone = progress?.call(
-          'Running "flutter test" in $cwd',
+    void Function(String)? stdout,
+    void Function(String)? stderr,
+  }) {
+    return _runCommand(
+      cmd: (cwd) {
+        void noop(String? _) {}
+        stdout?.call('Running "flutter test" in $cwd...\n');
+        return _flutterTest(
+          cwd: cwd,
+          stdout: stdout ?? noop,
+          stderr: stderr ?? noop,
         );
-        try {
-          final result = await _Cmd.run(
-            'flutter',
-            ['test'],
-            workingDirectory: cwd,
-          );
-          return result;
-        } catch (_) {
-          rethrow;
-        } finally {
-          installDone?.call();
-        }
       },
       cwd: cwd,
       recursive: recursive,
     );
   }
+}
+
+/// Run a command on directories with a `pubspec.yaml`.
+Future<void> _runCommand<T>({
+  required Future<T> Function(String cwd) cmd,
```

Comment:
Nit: maybe have a type def for this?

Reply:
Why do you feel it's necessary? This is a private API and by providing a typedef developers would have to jump to the typedef to understand the function signature.

Reply:
Just for the sake of DRY, I've noticed this is repeated along the file.

Either way this is just a nit, I am fine either way

Reply:
Where else do you see it repeated? I only see `Future<T> Function(String cwd)` in one spot.

Reply:
Yeah, this isn't repeated, I guess I got into the flow because of the other comments